### DISCUSSION TITLE
Custom Database Roles

### DISCUSSION THREAD

**Discussion by ozum** [supabase/supabase#11799] [Feature Requests]
# Feature request

## Is your feature request related to a problem? Please describe.

In a multi-tenant app, I would have 5-10 user levels, which would have different column level authorization. For example, a tenant would have such user groups: "client", "employee", "manager", "director", "approver", "admin", etc. Each level has different select, insert and update authorization for columns. For example, a "client" can see product names, but not prices. An "employee" can see product prices, and update product features, but can not update prices. An "approver" can only see the transaction date, and sale price, and only update the `is_approved` field.

## Describe the solution you'd like

I would like Supabase to support custom database users, so I can group app users under several database users. So I can combine app user identity (for Row Level Security) with a database user identity (for Column Level Security).

## Describe alternatives you've considered

Database views or database functions are the current supabase practice to implement column-level security (#11670), but it has several major drawbacks. For example, having 5 user levels for 3 operations (`SELECT`, `INSERT`, `UPDATE`)

* 3 * 5 = 15 procedures for each table, which means hundreds of procedures for 20-30 tables.
* Otherwise, 3 * 5 = 15 views for each table resulting in hundreds of views.
* Each table would have 15 REST and API endpoints.
* Having lots of endpoints for the same purpose and table but for different user levels, the API would be a mess.
* It is a maintenance nightmare to drop and recreate those 15 views or 15 procedures each time the related table changes.

If Supabase has supported custom database roles, we can just create 5 database users for user levels and group app users within those roles to apply column-level security elegantly.

## Additional context

As discussed at #3107, #3129, #3232, and #11670, there is a valid use case for having custom database role support in Supabase besides the hardcoded `authenticated`, `anon`, and `service_role`. postgREST already supports [Hybrid User-Group Roles](https://postgrest.org/en/stable/auth.html#hybrid-user-group-roles) for a long time.

On Sep 2021,  Supabase seemed to have this feature:

> @steve-chavez [mentioned](https://github.com/supabase/supabase/discussions/3107#discussioncomment-1292883): Having said that, [different roles for grouping many web users](https://postgrest.org/en/v8.0/auth.html#web-users-sharing-role) is feasible. Right now Supabase only supports the authenticated and anon groups but we'll later support adding your own custom groups(this already works in PostgREST but needs additional work for Kong).

> @awalias [mentioned](https://github.com/supabase/supabase/discussions/3232#discussioncomment-1343550): .... our API gateway does not yet currently support custom "postgres level roles", this is being worked on and should be available in the coming months.

**Comment by kiwicopple** [comment]
Hi @ozum - Supabase does support Custom Roles already, although you'd need to create them from the SQL Editor yourself (we don't provide a UI for them yet):

```
create role "my_custom_user";
```

From there, if you want them to have access through the API then you need to mint a JWT for that user with the "role" field matching the name of the role. You can see a full example here:

https://supabase.com/docs/guides/self-hosting#api-keys

<img width="390" alt="image" src="https://user-images.githubusercontent.com/10214025/213701230-d1790a02-a1eb-47bb-afb0-5eaad8657395.png">


You can find your JWT Secret in the Settings (under "API")

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/10214025/213701442-00d347a3-e827-46c9-8714-428c5581f526.png">


Hopefully that helps! If you want to mind your JWTs dynamically, please make sure to do that server-side, so that you don't leak your JWT secret. I'll convert this to a Discussion for now, but if there are specific blockers to this approach, feel free to open a bug report / issue 👍

**Reply by ozum** [reply]
@kiwicopple, thanks that's awesome, although it is undocumented. Creating roles via SQL is OK for me, I already use SQL and migrations.

1) Could you please elaborate on how this fits/integrate with [Supabase Auth](https://supabase.com/docs/guides/auth/overview#how-it-works) and `@supabase/supabase-js`?

2) Does it work with "social logins", "password reminder", "magic link" etc?

3) How can I use it with nuxt supabase login workflow?
	For example, I'm currently using Nuxt Supabase described in supabase docs [here](https://supabase.com/docs/guides/getting-started/tutorials/with-nuxt-3)

```ts
    const { data, error } = await auth.signInWithPassword({
      email: email.value,
      password: password.value,
    });
```

4) Does Supabase and/or postGREST in Supabase works as expected if it sees a custom role in JWT?
5) Document you referred is for self-hosting. Will it work in the Supabase Pro plan?

EDIT: Sorry for the many questions, I'm wrestling with this issue for a week...

Many thanks,

**Reply by jodator** [reply]
@kiwicopple Is this feature enabled for hosted supabase or only for self-hosted? I'm trying to enable custom role and I get

```json
{
  "message": "Invalid API key",
  "hint": "Double check your Supabase `anon` or `service_role` API key."
}
```

I double checked if I can recreate `service_key` or `anon` using the same steps and I get 1:1 match. However, if I set different role, like:

```json
{
  "iss": "supabase",
  "ref": "yourprojectrefwink",
  "role": "my_custom_role",
  "exp": 2005475488,
  "iat": 1689856288
}
```

The SDK functions returns the error above.

**Comment by steve-chavez** [comment]
@ozum Check https://github.com/supabase/supabase/discussions/11948#discussioncomment-4802137, a user managed to get custom roles working.

**Reply by ozum** [reply]
@steve-chavez, thank you for following up and pointing the message. Really helped.

**Comment by ruggi99** [comment]
@ozum started a new PR to move the first step to bring custom roles and column permission to Supabase: supabase/postgres-meta#491.
Hope they review and eventually merge soon.
Next steps would be:
- Add a role via Supabase UI and grant access to `authenticator` role
- Manage column permission via Supabase UI

**Comment by rienheuver** [comment]
Great to here that additional roles are possible! Is there any way of using this in local development? The dockers are served from `supabase start` and I haven't yet found a way to add a `.env` to that or to influence the `kong.yml`. For that matter, I haven't been able to set custom anon/service_role keys for local development either.
